src/test_array_builder.cc

```diff
@@ -381,35 +428,97 @@ Query::Status writeData(Context ctx, std::string array_name, tiledb_datatype_t d
-      values->push_back(1);
-      values->push_back(1);
-      query.set_coordinates(*values);
-      buffers.emplace_back(nullptr, std::move(values));
+      std::shared_ptr<std::vector<uint64_t>> d_row = std::make_shared<std::vector<uint64_t>>();
+      std::shared_ptr<std::vector<uint64_t>> d_col = std::make_shared<std::vector<uint64_t>>();
+      d_row->push_back(1);
+      d_col->push_back(1);
+      query.set_buffer("rows",*d_row)
+            .set_buffer("cols",*d_col);
+      buffers.emplace_back(nullptr, std::move(d_row));
+      buffers.emplace_back(nullptr, std::move(d_col));
       break;
     }
     case TILEDB_FLOAT32: {
-      std::shared_ptr<std::vector<float>> values = std::make_shared<std::vector<float>>();
-      values->push_back(1);
-      values->push_back(1);
-      query.set_coordinates(*values);
-      buffers.emplace_back(nullptr, std::move(values));
+      std::shared_ptr<std::vector<float>> d_row = std::make_shared<std::vector<float>>();
+      std::shared_ptr<std::vector<float>> d_col = std::make_shared<std::vector<float>>();
+      d_row->push_back(1);
+      d_col->push_back(1);
+      query.set_buffer("rows",*d_row)
+            .set_buffer("cols",*d_col);
+      buffers.emplace_back(nullptr, std::move(d_row));
+      buffers.emplace_back(nullptr, std::move(d_col));
       break;
     }
     case TILEDB_FLOAT64: {
-      std::shared_ptr<std::vector<double>> values = std::make_shared<std::vector<double>>();
-      values->push_back(1);
-      values->push_back(1);
-      query.set_coordinates(*values);
-      buffers.emplace_back(nullptr, std::move(values));
+      std::shared_ptr<std::vector<double>> d_row = std::make_shared<std::vector<double>>();
+      std::shared_ptr<std::vector<double>> d_col = std::make_shared<std::vector<double>>();
+      d_row->push_back(1);
+      d_col->push_back(1);
+      query.set_buffer("rows",*d_row)
+            .set_buffer("cols",*d_col);
+      buffers.emplace_back(nullptr, std::move(d_row));
+      buffers.emplace_back(nullptr, std::move(d_col));
+      break;
+    }
+    case TILEDB_CHAR: {
+      std::shared_ptr<std::vector<char>> d_row = std::make_shared<std::vector<char>>();
+      std::shared_ptr<std::vector<char>> d_col = std::make_shared<std::vector<char>>();
+      d_row->push_back('1');
+      d_col->push_back('1');
+      query.set_buffer("rows", *d_row)
+            .set_buffer("cols", *d_col);
+      buffers.emplace_back(nullptr, std::move(d_row));
+      buffers.emplace_back(nullptr, std::move(d_col));
+      break;
+    }
+    case TILEDB_STRING_ASCII:
+    case TILEDB_STRING_UTF8:{
+      std::shared_ptr<std::vector<std::string>> d_row = std::make_shared<std::vector<std::string>>();
+      std::shared_ptr<std::vector<std::string>> d_col = std::make_shared<std::vector<std::string>>();
+      std::string val = "1";
+      d_row->push_back(val);
+      d_col->push_back(val);
+      query.set_buffer("rows",*offsets,*d_row)
```

Comment:
Use a different vector<uint64_t> instances for each dimension. Something like `offsets_row` and `offsets_col` is that I would use.

Additionally, please populate these buffers. The `addDataToQuery` adds a single `0` element to the offsets it creates.

Reply:
I got a little bit confused by this. Currently I create one `offset` vector that is instantiated with `0` value `offsets->push_back(0)`. Do you recommend having 2 such vectors for different dimensions right? In that case though what is the use of `buffers` in the previous comment? I will have to define these variables. Also `buffers` is only declared. Of course I can assign the `offset` pointer to the first element of the `buffers` pair and use it from there. But again since `emplace_back` happens last I need to have the `offsets` vector. Could you explain it a little bit further?

Reply:
>Do you recommend having 2 such vectors for different dimensions right?
Yes

I'm saying that we don't need the existing `offsets` variable at the scope of `buffers`. The `buffers` exists to store a pair of offsets and attribute data. So I'm looking for:
```
case TILEDB_STRING_ASCII:
case TILEDB_STRING_UTF8:{
  std::shared_ptr<std::vector<std::string>> d_row = std::make_shared<std::vector<std::string>>();
  std::shared_ptr<std::vector<std::string>> d_col = std::make_shared<std::vector<std::string>>();
  std::string val = "1";
  d_row->push_back(val);
  d_col->push_back(val);

  std::unique_ptr<std::vector<uint64_t>> offsets_row =
    std::unique_ptr<std::vector<uint64_t>>(new std::vector<uint64_t>);
  std::unique_ptr<std::vector<uint64_t>> offsets_col =
    std::unique_ptr<std::vector<uint64_t>>(new std::vector<uint64_t>);
  offsets_row->emplace_back(0);
  offsets_col->emplace_back(0);

  query.set_buffer("rows", *offsets_row, *d_row)
       .set_buffer("cols", *offsets_col, *d_col);
  buffers.emplace_back(std::move(offsets_row), std::move(d_row));
  buffers.emplace_back(std::move(offsets_col), std::move(d_col));
  break;
}
```

Reply:
Perfect!!! exactly what I had in mind thnx